### ISSUE TITLE
Multi-gpu training

### ISSUE THREAD

**Issue by JohnGiorgi** [JohnGiorgi/t2t#9]
In [SimCLR](https://arxiv.org/abs/2002.05709), the number of _negatives_ per some _positive_ pair is taken to be `2 * (N - 1)` (all examples in a minibatch of size `N` that don't belong to that positive pair), and they find ([as other works before them](https://papers.nips.cc/paper/6200-improved-deep-metric-learning-with-multi-class-n-pair-loss-objective)) that the bigger the batch size, the larger the number of negatives, and the better the learned representations.

`DataParallel` and `DistributedDataParallel` divide up a mini-batch, send each partition of examples to a GPU, compute gradients, and then average these gradients before backpropagating. But this means that each GPU is computing a loss with `N/n_gpus` examples and, therefore,  `2 * (N/n_gpus - 1)` negatives per positive pair.

We need to figure out how to use multi-GPU training, preferably with `DistributedDataParallel`, in a way that avoids this "issue", I.e. that allows us to use multiple GPUs while maintaining `2 * (N - 1)` negatives per positive pair.

This is high priority as we will need multiple GPUs to scale the number of negatives in order to maximize performance (and make training on millions of documents feasible).
